lib/ecto/query/planner.ex

```diff
@@ -1913,8 +1914,13 @@ defmodule Ecto.Query.Planner do
   end
 
   defp type!(kind, query, expr, schema, field, allow_virtuals?) when is_atom(schema) do
+    only_writable? = kind in @write_only_kinds
+
     cond do
-      type = schema.__schema__(:type, field) ->
+      type = only_writable? && schema.__schema__(:writable_type, field) ->
+        type
+
+      type = !only_writable? && schema.__schema__(:type, field) ->
         type
 
       type = allow_virtuals? && schema.__schema__(:virtual_type, field) ->
```

Comment:
I was wondering if we could unify this somehow, so we don't have to perform three different queries, but I have no idea. Plus `:virtual_type` is public, so it is not like we can get rid of it.

Reply:
When I was playing around with this they don't really intersect well. Virtual types are only allowed in `type/2` so that's where that restriction comes in. And the writable restriction comes in with being part of an update.

I think it is probably cleaner to put it back into assert_update? We are already traversing the update keys there https://github.com/elixir-ecto/ecto/blob/50d81d8d35317de9f5091c4839cf3775c6786db9/lib/ecto/query/planner.ex#L2029